net/ip.c

```diff
@@ -0,0 +1,448 @@
+		if (nr != 2)
+			return M0_ERR(-EPROTO);
+		for (i = 0; i < ARRAY_SIZE(ip_autotm); ++i) {
+			if (ip_autotm[i] == 0) {
+				tmid = i;
+				/* To handle '*' wildchar as tmid*/
+				addr->nia_n.fmt_pvt.la.nla_autotm = true;
+				break;
+			}
+		}
+		if (i == ARRAY_SIZE(ip_autotm))
+			return M0_ERR(-EADDRNOTAVAIL);
+	} else
+		addr->nia_n.fmt_pvt.la.nla_autotm = false;
+
+	if (pid != 12345)
+		return M0_ERR(-EPROTO);
+	/*
+	* Deterministically combine portal and tmid into a unique 16-bit port
+	* number (greater than 1024). Tricky.
+	*
+	* Port number is, in binary: tttttttttt1ppppp, that is, 10 bits of tmid
+	* (which must be less than 1024), followed by a set bit (guaranteeing
+	* that the port is not reserved), followed by 5 bits of (portal - 30),
+	* so that portal must be in the range 30..61.
+	*
+	if (tmid >= 1024 || (portal - 30) >= 32)
+		return M0_ERR_INFO(-EPROTO,
+			"portal: %u, tmid: %u", portal, tmid);
+	*/
+
+	addr->nia_n.fmt_pvt.la.nla_portal =  portal;
+	if (portal < 30)
+		portal = 30 + portal;
+
+	portnum  = tmid | (1 << 10) | ((portal - 30) << 11);
+	M0_ASSERT(portnum < M0_NET_IP_PORT_MAX);
+	snprintf(port, ARRAY_SIZE(port), "%d", (int)portnum);
+	ip_autotm[tmid] = 1;
+
+	addr->nia_n.nip_format = M0_NET_IP_LNET_FORMAT;
+	inet_pton(AF_INET, node, &addr->nia_n.ip_n.sn[0]);
+	addr->nia_n.fmt_pvt.la.nla_tmid = tmid;
+	addr->nia_n.nip_port = (uint16_t)portnum;
+	M0_ASSERT(strlen(name) < ARRAY_SIZE(addr->nia_p));
+	strcpy(addr->nia_p, name);
+
+	return M0_RC(0);
+}
+
+/**
+ * Compare ipv4 address in network byte order.
+ */
+static bool m0_net_ip_v4_cmp(uint32_t *a1, uint32_t *a2)
+{
+	return (a1[0] == a2[0]);
+}
+
+/**
+ * Compare ipv6 address in network byte order.
+ */
+static bool m0_net_ip_v6_cmp(uint64_t *a1, uint64_t *a2)
+{
+	return (a1[0] == a2[0] && a1[1] == a2[1]);
+}
+
+/**
+ * Compare lnet address format specific parameters.
+ */
+static bool m0_net_ip_la_cmp(struct m0_net_ip_addr *a1,
+			     struct m0_net_ip_addr *a2)
+{
+	struct m0_net_ip_lnet_addr *la1;
+	struct m0_net_ip_lnet_addr *la2;
+
+	M0_PRE(a1 != NULL && a2 != NULL);
+	la1 = &a1->nia_n.fmt_pvt.la;
+	la2 = &a2->nia_n.fmt_pvt.la;
+
+	return (la1->nla_type   == la2->nla_type   &&
+		la1->nla_portal == la2->nla_portal &&
+		la1->nla_tmid   == la2->nla_tmid   &&
+		la1->nla_autotm == la2->nla_autotm &&
+		m0_net_ip_v4_cmp(&a1->nia_n.ip_n.sn[0], &a2->nia_n.ip_n.sn[0]));
+}
+
+/**
+ * Compare inet address format specific parameters.
+ */
+static bool m0_net_ip_ia_cmp(struct m0_net_ip_addr *a1,
+			     struct m0_net_ip_addr *a2)
+{
+	struct m0_net_ip_inet_addr *ia1;
+	struct m0_net_ip_inet_addr *ia2;
+
+	M0_PRE(a1 != NULL && a2 != NULL);
+	ia1 = &a1->nia_n.fmt_pvt.ia;
+	ia2 = &a2->nia_n.fmt_pvt.ia;
+
+	return (ia1->nia_family == ia2->nia_family &&
+		ia1->nia_type   == ia2->nia_type &&
+		ia1->nia_family == M0_NET_IP_AF_INET ?
+		m0_net_ip_v4_cmp(&a1->nia_n.ip_n.sn[0], &a2->nia_n.ip_n.sn[0]) :
+		m0_net_ip_v6_cmp(&a1->nia_n.ip_n.ln[0], &a2->nia_n.ip_n.ln[0]));
+}
```

Comment:
`_cmp` suffix is usually used for 3-way comparisons. For equality checks we have `_eq` (for example, m0_streq, m0_buf_eq).

Reply:
Renamed functions